### PR TITLE
InterceptUrlMapRule by exact match on both method and URI

### LINKED ISSUE #1511: InterceptUrlMapRule: Configuration Matching Bug
### Steps to Reproduce

1. Enable the security feature
```
# build.gradle
dependencies {
    annotationProcessor "io.micronaut:micronaut-security"
    implementation "io.micronaut:micronaut-security"
    ...
}
```

```
# application.yml
micronaut:
  security:
    enabled: true
```

2. Add an `intercept-url-map` in the app config for a specific route with a different behaviour depending on the `httpMethod`

```
# application.yml
micronaut:
  security:
    enabled: true
    intercept-url-map:
      - pattern: /v1/sessions/**
        access:
          - isAuthenticated()
      - pattern: /v1/sessions/**
        httpMethod: OPTIONS
        access:
          - isAnonymous()
```

### Expected Behaviour

`GET /v1/sessions/ABC` needs authentication
`POST /v1/sessions/` needs authentication
`PUT /v1/sessions/ABC` needs authentication
`OPTIONS /v1/sessions/` needs NO authentication

### Actual Behaviour

`GET /v1/sessions/ABC` needs authentication
`POST /v1/sessions/` needs authentication
`PUT /v1/sessions/ABC` needs authentication
`OPTIONS /v1/sessions/` needs authentication

### Additional Information

This seems to be an issue with `io.micronaut.security.rules.InterceptUrlMapRule`, specifically this part:

```
        Optional<InterceptUrlMapPattern> matchedPattern = getPatternList()
                .stream()
                .filter(p -> p.getHttpMethod().map(method -> method.equals(httpMethod)).orElse(true))
                .filter(p -> pathMatcher.matches(p.getPattern(), path))
                .findFirst();
```

The assumption here seems to be, that patterns are unique. In my case the patterns are not unique, but the combination of patterns and httpMethods are.

### Environment Information

- **Operating System**: macOS Mojave 10.14.4
- **Micronaut Version:** 1.1.0.RC2 (Issue persists with 1.1.0.BUILD-SNAPSHOT)
- **JDK Version:** openjdk version "11.0.2" 2019-01-15

### PR DESCRIPTION
**Pull request by saw303**
This PR fixes #1511 by first trying to find  url rules that **exactly match** the http **request method** and the **uri pattern**. 
If there is no match the code falls back to the original behaviour by finding by uri pattern.

### CODE REVIEW

File: security/src/main/java/io/micronaut/security/rules/InterceptUrlMapRule.java [hunk 1]
```diff
@@ -83,12 +87,22 @@ public SecurityRuleResult check(HttpRequest request, @Nullable RouteMatch routeM
         final String path = request.getUri().getPath();
         final HttpMethod httpMethod = request.getMethod();
 
+        Predicate<InterceptUrlMapPattern> exactMatch = p -> pathMatcher.matches(p.getPattern(), path) && p.getHttpMethod().isPresent() && httpMethod.equals(p.getHttpMethod().get());
+        Predicate<InterceptUrlMapPattern> uriPatternMatch = p -> pathMatcher.matches(p.getPattern(), path) && p.getHttpMethod().map(method -> method.equals(httpMethod)).orElse(true);
```
Comment (jameskleeh) [inline security/src/main/java/io/micronaut/security/rules/InterceptUrlMapRule.java]: Shouldn't this just check that `!p.getHttpMethod().isPresent()`

**Comment by jameskleeh** [comment]
This needs to target master. Only bugfixes in the 1.1.x branch

File: security/src/main/java/io/micronaut/security/rules/InterceptUrlMapRule.java (hunk 1 above)
Comment (saw303) [reply security/src/main/java/io/micronaut/security/rules/InterceptUrlMapRule.java]: Good catch. Changed the predicate

**Comment by jameskleeh** [comment]
@saw303 Can you add some information to the documentation? https://github.com/micronaut-projects/micronaut-core/blob/master/src/main/docs/guide/security/securityRule/interceptUrlMap.adoc

**Comment by saw303** [comment]
@jameskleeh is that understandable? My writing in English is not the best :(

**Comment by jameskleeh** [comment]
Thanks for your contribution!

**Comment by saw303** [comment]
@jameskleeh you are welcome

**Pull request merged**